### ISSUE TITLE
jvmci compiler with jdk15 fails due to missing classes

### ISSUE THREAD

**Issue by pmlopes** [oracle/graal#2881]
If we try to run a graaljs application with stock jdk15, using the jvmci updated module, the vm fails to start with:

```
org.graalvm.polyglot.PolyglotException: java.lang.NoClassDefFoundError: jdk/vm/ci/sparc/SPARC
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.sparc.SPARCHotSpotBackendFactory.getArchitecture(SPARCHotSpotBackendFactory.java:86)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.CompilerConfigurationFactory$DefaultBackendMap.<init>(CompilerConfigurationFactory.java:130)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.CompilerConfigurationFactory.createBackendMap(CompilerConfigurationFactory.java:107)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalRuntime.<init>(HotSpotGraalRuntime.java:200)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompilerFactory.createCompiler(HotSpotGraalCompilerFactory.java:156)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompilerFactory.createCompiler(HotSpotGraalCompilerFactory.java:134)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompilerFactory.createCompiler(HotSpotGraalCompilerFactory.java:52)
        at jdk.internal.vm.ci/jdk.vm.ci.hotspot.HotSpotJVMCIRuntime.getCompiler(HotSpotJVMCIRuntime.java:630)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.compiler.hotspot.HotSpotTruffleCompilerImpl.getCompiler(HotSpotTruffleCompilerImpl.java:144)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.compiler.hotspot.HotSpotTruffleCompilerImpl.create(HotSpotTruffleCompilerImpl.java:121)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.java.HotSpotTruffleRuntime.newTruffleCompiler(HotSpotTruffleRuntime.java:95)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.AbstractHotSpotTruffleRuntime.initializeTruffleCompiler(AbstractHotSpotTruffleRuntime.java:251)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.AbstractHotSpotTruffleRuntime.access$200(AbstractHotSpotTruffleRuntime.java:85)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.AbstractHotSpotTruffleRuntime$1.execute(AbstractHotSpotTruffleRuntime.java:217)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.BackgroundCompileQueue$RequestImpl.call(BackgroundCompileQueue.java:236)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base/java.lang.Thread.run(Thread.java:832)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.BackgroundCompileQueue$TruffleCompilerThreadFactory$1.run(BackgroundCompileQueue.java:282)
Caused by: java.lang.ClassNotFoundException: jdk.vm.ci.sparc.SPARC
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:606)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        ... 20 more
Caused by: Attached Guest Language Frames (0)
Internal GraalVM error, please report at https://github.com/oracle/graal/issues/.
Failed in deploying verticle caused by org.graalvm.polyglot.PolyglotException: java.lang.NoClassDefFoundError: jdk/vm/ci/sparc/SPARC
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.sparc.SPARCHotSpotBackendFactory.getArchitecture(SPARCHotSpotBackendFactory.java:86)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.CompilerConfigurationFactory$DefaultBackendMap.<init>(CompilerConfigurationFactory.java:130)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.CompilerConfigurationFactory.createBackendMap(CompilerConfigurationFactory.java:107)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalRuntime.<init>(HotSpotGraalRuntime.java:200)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompilerFactory.createCompiler(HotSpotGraalCompilerFactory.java:156)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompilerFactory.createCompiler(HotSpotGraalCompilerFactory.java:134)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompilerFactory.createCompiler(HotSpotGraalCompilerFactory.java:52)
        at jdk.internal.vm.ci/jdk.vm.ci.hotspot.HotSpotJVMCIRuntime.getCompiler(HotSpotJVMCIRuntime.java:630)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.compiler.hotspot.HotSpotTruffleCompilerImpl.getCompiler(HotSpotTruffleCompilerImpl.java:144)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.compiler.hotspot.HotSpotTruffleCompilerImpl.create(HotSpotTruffleCompilerImpl.java:121)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.java.HotSpotTruffleRuntime.newTruffleCompiler(HotSpotTruffleRuntime.java:95)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.AbstractHotSpotTruffleRuntime.initializeTruffleCompiler(AbstractHotSpotTruffleRuntime.java:251)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.AbstractHotSpotTruffleRuntime.access$200(AbstractHotSpotTruffleRuntime.java:85)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.AbstractHotSpotTruffleRuntime$1.execute(AbstractHotSpotTruffleRuntime.java:217)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.BackgroundCompileQueue$RequestImpl.call(BackgroundCompileQueue.java:236)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base/java.lang.Thread.run(Thread.java:832)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.BackgroundCompileQueue$TruffleCompilerThreadFactory$1.run(BackgroundCompileQueue.java:282)
Caused by: java.lang.ClassNotFoundException: jdk.vm.ci.sparc.SPARC
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:606)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        ... 20 more
Caused by: Attached Guest Language Frames (0)
Internal GraalVM error, please report at https://github.com/oracle/graal/issues/.
org.graalvm.polyglot.PolyglotException: java.lang.NoClassDefFoundError: jdk/vm/ci/sparc/SPARC
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.sparc.SPARCHotSpotBackendFactory.getArchitecture(SPARCHotSpotBackendFactory.java:86)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.CompilerConfigurationFactory$DefaultBackendMap.<init>(CompilerConfigurationFactory.java:130)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.CompilerConfigurationFactory.createBackendMap(CompilerConfigurationFactory.java:107)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalRuntime.<init>(HotSpotGraalRuntime.java:200)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompilerFactory.createCompiler(HotSpotGraalCompilerFactory.java:156)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompilerFactory.createCompiler(HotSpotGraalCompilerFactory.java:134)
        at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompilerFactory.createCompiler(HotSpotGraalCompilerFactory.java:52)
        at jdk.internal.vm.ci/jdk.vm.ci.hotspot.HotSpotJVMCIRuntime.getCompiler(HotSpotJVMCIRuntime.java:630)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.compiler.hotspot.HotSpotTruffleCompilerImpl.getCompiler(HotSpotTruffleCompilerImpl.java:144)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.compiler.hotspot.HotSpotTruffleCompilerImpl.create(HotSpotTruffleCompilerImpl.java:121)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.java.HotSpotTruffleRuntime.newTruffleCompiler(HotSpotTruffleRuntime.java:95)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.AbstractHotSpotTruffleRuntime.initializeTruffleCompiler(AbstractHotSpotTruffleRuntime.java:251)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.AbstractHotSpotTruffleRuntime.access$200(AbstractHotSpotTruffleRuntime.java:85)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.hotspot.AbstractHotSpotTruffleRuntime$1.execute(AbstractHotSpotTruffleRuntime.java:217)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.BackgroundCompileQueue$RequestImpl.call(BackgroundCompileQueue.java:236)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
        at java.base/java.lang.Thread.run(Thread.java:832)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.BackgroundCompileQueue$TruffleCompilerThreadFactory$1.run(BackgroundCompileQueue.java:282)
Caused by: java.lang.ClassNotFoundException: jdk.vm.ci.sparc.SPARC
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:606)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        ... 20 more
Caused by: Attached Guest Language Frames (0)
Internal GraalVM error, please report at https://github.com/oracle/graal/issues/.
```

This seems to be related to JEP 381.

**Comment by dougxc** [comment]
This requires an mx fix. Can you please test this patch to confirm it fixes the problem on your end:
```
diff --git a/mx.py b/mx.py
index d0bfe33..b7aface 100755
--- a/mx.py
+++ b/mx.py
@@ -4381,6 +4381,7 @@ def _remove_unsatisfied_deps():
     A reason may be the name of another removed dependency.
     """
     removedDeps = OrderedDict()
+    default_jdk = get_jdk(tag=DEFAULT_JDK_TAG)
 
     def visit(dep, edge):
         if dep.isLibrary():
@@ -4445,6 +4446,11 @@ def _remove_unsatisfied_deps():
                     logv('[{} was removed from distribution {}]'.format(distDep, dist))
                     dist.removeDependency(distDep)
                     distRemovedDeps.append(distDep)
+                elif distDep.isJavaProject():
+                    if default_jdk.javaCompliance not in distDep.javaCompliance:
+                        logv('[{} was removed from distribution {}]'.format(distDep, dist))
+                        dist.removeDependency(distDep)
+                        distRemovedDeps.append(distDep)
 
             if discard(dist):
                 note_removal(dist, 'distribution {} was removed as all its dependencies were removed'.format(dist),
```

**Comment by dougxc** [comment]
Should be fixed by https://github.com/graalvm/mx/commit/2d41ec7beb2d673d694ba5cf11fe04aa68e24a73 - please re-open if not.

**Comment by stevenschlansker** [comment]
We're running openjdk 15.0.1 and graal 20.3 and are suffering this problem.
I don't see newer versions of the JAR files to link against -- we don't run `mx` at all, how can we incorporate the fix?

**Comment by dougxc** [comment]
This was re-fixed in 76d0c303763cbd1d98ffee722ef7644819143aa3 which did not make it into 20.3.
Where are you getting the Graal JAR files?

**Comment by stevenschlansker** [comment]
We link Graal in as a library from Maven Central:
https://search.maven.org/artifact/org.graalvm.compiler/compiler/20.3.0/jar

**Comment by dougxc** [comment]
Using the maven jars with any JDK for which there is not a GraalVM release will always run this risk.

**Comment by stevenschlansker** [comment]
Yes, this probably isn't the most common configuration, and as open source users it broke and we got to keep both pieces :)
All the same, hoping for the latest Graal GA to work with latest OpenJDK GA (especially given the new rolling release schedule) doesn't seem like that far of a reach.  Looking forward to the fix hopefully landing in 20.3.1, we'll try to patch the jar ourselves in the meantime...
